posts/01-effective-rust-canisters.html.pm

```diff
@@ -442,6 +442,355 @@ type CreateEntityResult = variant {
 }
 }
 
+◊section{
+◊section-title["optimization"]{Optimization}
+◊subsection-title["cycle-consumption"]{Reducing cycle consumption}
+◊p{
+  The first step towards an optimized system is profiling.
+}
+
+◊advice["performance-counter"]{Measure the number of instructions your endpoints consume.}
+
+◊p{
+  The ◊a[#:href "https://docs.rs/ic-cdk/0.5.2/ic_cdk/api/call/fn.performance_counter.html"]{◊code{performance_counter}} API will tell you the number of ◊em{instructions} your code consumed since the last ◊a[#:href "https://internetcomputer.org/docs/current/references/ic-interface-spec/#entry-points"]{entry point}.
+  Instructions are the internal currency of the IC runtime: the system defines all its limits in terms of instructions.
```

Comment:
Shall we explain it in terms of WebAssembly instructions? It's basically what it is, unless it's a sys call...

Reply:
> Shall we explain it in terms of WebAssembly instructions?

Done. 

>  It's basically what it is

Almost, but not really: there is a lookup table that maps Wasm instructions to IC instructions. Most of the Wasm instructions have a weight of 1, some have a weight of 0. I can imagine we can change the mapping at some point.